File: backend/chunking-rag/src/Reasoning/prompt_builder.ts
/**
 * 提示词构建器
 * 核心要求 4: 边界严控（拒答）- 构建严格的提示词，确保模型只基于上下文回答
 */

import { ContextBlock } from './types.js';

/**
 * 系统提示词模板
 */
const SYSTEM_PROMPT = `你是一个严格的技术文档问答助手。

【核心规则】
1. 仅根据下方提供的 Context 回答，严禁使用任何外部知识或猜测。
2. 每个事实性陈述句末必须标注来源，格式为 [n]（n 为对应 Chunk 的 ID）。
3. 如果 Context 中不包含回答所需信息，直接回复："根据现有文档无法回答此问题。"
4. 如果 Context 标注了"内容已截断"，可说明信息不足并建议查阅原文。
5. 不得合并多个 Chunk 的内容进行推断，每条引用必须有直接支撑。
6. 保持回答简洁准确，不要添加冗余解释。

【引用规则】
- 每个句子的引用必须是直接的、一一对应的
- 避免泛泛而谈，每个观点都需要具体引用支撑
- 数字、版本号、配置项等关键信息必须标注来源`;

/**
 * 用户提示词模板
 */
const USER_PROMPT_TEMPLATE = `【Context】
{context}

---

【问题】
{query}

---

【回答要求】
请严格基于 Context 回答，每个事实陈述后标注引用。`;

/**
 * 拒答提示词（当检索得分过低时）
 */
const REJECTION_PROMPT = `根据现有文档无法回答此问题。

提示：当前检索得分（{maxScore:.2f}）低于系统阈值，无法确保回答准确性。
建议：
1. 尝试重新表述问题
2. 上传更多相关文档
3. 确认文档中确实包含此信息`;

/**
 * 无 LLM 模式提示词
 */
const NO_LLM_PROMPT = `【问题】
{query}

【检索到的相关文档】
{context}

---
根据上述检索结果，以下是相关信息汇总：

{summary}`;

/**
 * 提示词构建器
 */
export class PromptBuilder {
  private systemPrompt: string;

  constructor(systemPrompt?: string) {
    this.systemPrompt = systemPrompt || SYSTEM_PROMPT;
  }

  /**
   * 构建完整的推理提示词
   */
  build(
    query: string,
    contextBlocks: ContextBlock[],
    contextTruncated: boolean = false
  ): { system: string; user: string } {
    const context = this.formatContext(contextBlocks);
    
    let extendedSystemPrompt = this.systemPrompt;
    
    // 如果上下文被截断，添加警告
    if (contextTruncated) {
      extendedSystemPrompt += '\n\n⚠️ 警告：以下 Context 因长度限制被截断，可能不包含完整信息。';
    }

    const userPrompt = USER_PROMPT_TEMPLATE
      .replace('{context}', context)
      .replace('{query}', query);

    return {
      system: extendedSystemPrompt,
      user: userPrompt,
    };
  }

  /**
   * 构建拒答提示词
   */
  buildRejection(maxScore: number): { system: string; user: string } {
    return {
      system: this.systemPrompt,
      user: REJECTION_PROMPT.replace('{maxScore}', maxScore.toFixed(2)),
    };
  }

  /**
   * 构建无 LLM 模式提示词（仅返回检索结果）
   */
  buildNoLLM(
    query: string,
    contextBlocks: ContextBlock[]
  ): { system: string; user: string } {
    const context = this.formatContext(contextBlocks);
    const summary = this.summarizeResults(contextBlocks);

    return {
      system: '你是一个信息检索助手，负责汇总检索结果。',
      user: NO_LLM_PROMPT
        .replace('{query}', query)
        .replace('{context}', context)
        .replace('{summary}', summary),
    };
  }

  /**
   * 格式化上下文块
   */
  private formatContext(blocks: ContextBlock[]): string {
    if (blocks.length === 0) {
      return '（无相关文档）';
    }

    return blocks
      .map(block => {
        let formatted = `[ID: ${block.id}, Source: ${block.source}]\n`;
        formatted += block.content;
        if (block.isTruncated) {
          formatted += '\n[此段内容已截断，建议查阅原文]';
        }
        return formatted;
      })
      .join('\n\n---\n\n');
  }

  /**
   * 汇总检索结果（无 LLM 模式）
   */
  private summarizeResults(blocks: ContextBlock[]): string {
    return blocks
      .map((block, idx) => `${idx + 1}. ${block.content.substring(0, 200)}${block.content.length > 200 ? '...' : ''}`)
      .join('\n');
  }

  /**
   * 从 LLM 回答中提取引用 ID
   */
  extractCitationIds(answer: string): number[] {
    const regex = /\[(\d+)\]/g;
    const ids: number[] = [];
    let match;

    while ((match = regex.exec(answer)) !== null) {
      const id = parseInt(match[1], 10);
      if (!ids.includes(id)) {
        ids.push(id);
      }
    }

    return ids;
  }

  /**
   * 构建用于流式生成的单条消息
   */
  buildStreamMessage(
    query: string,
    contextBlocks: ContextBlock[],
    contextTruncated: boolean = false
  ): string {
    const context = this.formatContext(contextBlocks);
    
    let prompt = `【Context】\n${context}\n\n`;
    
    if (contextTruncated) {
      prompt += `⚠️ 注意：Context 因长度限制被截断，以下回答可能不完整。\n\n`;
    }
    
    prompt += `【问题】${query}\n\n`;
    prompt += `请严格基于 Context 回答，每个事实陈述后标注 [引用ID]。`;

    return prompt;
  }
}

/**
 * 创建提示词构建器
 */
export function createPromptBuilder(systemPrompt?: string): PromptBuilder {
  return new PromptBuilder(systemPrompt);
}

File: backend/reasoning/prompt_builder.py
"""
提示词构建器
核心要求 4: 边界严控（拒答）- 构建严格的提示词，确保模型只基于上下文回答

v2 变更：
  - 提示词模板改由 prompts/prompts.yaml 统一维护
  - 通过 config_loader.load_prompts_config() 加载（带 lru_cache）
  - 保留模块级常量（SYSTEM_PROMPT 等）作为向后兼容别名，首次访问时懒加载
"""

from __future__ import annotations
import re
from typing import List, Optional, Tuple
from ._types import ContextBlock


# ============================================================
# 懒加载模板（向后兼容，直接 import 常量仍可使用）
# ============================================================

def _get_prompts():
    """懒加载 PromptsConfig，避免模块导入时触发文件 IO"""
    from .config_loader import load_prompts_config
    return load_prompts_config()


# 向后兼容别名（首次访问时从 prompts.yaml 加载）
class _LazyStr(str):
    """惰性字符串代理，允许 `SYSTEM_PROMPT` 等常量在首次使用时才读取配置。"""
    pass


def __getattr__(name: str):
    """模块级 __getattr__：当常量被直接 import 时，从配置文件读取。"""
    _alias_map = {
        'SYSTEM_PROMPT':       lambda p: p.system_prompt,
        'USER_PROMPT_TEMPLATE': lambda p: p.user_prompt_template,
        'REJECTION_PROMPT':    lambda p: p.rejection_prompt,
        'NO_LLM_PROMPT':       lambda p: p.no_llm_prompt,
    }
    if name in _alias_map:
        return _alias_map[name](_get_prompts())
    raise AttributeError(f"module {__name__!r} has no attribute {name!r}")


# ============================================================
# PromptBuilder 类
# ============================================================

class PromptBuilder:
    """
    提示词构建器

    参数：
        system_prompt: 自定义系统提示词字符串（传 None 时从 prompts.yaml 读取）
        prompts_file:  自定义 prompts.yaml 路径（None = 使用默认路径）
    """

    def __init__(
        self,
        system_prompt: Optional[str] = None,
        prompts_file: Optional[str] = None,
    ):
        self._prompts_file = prompts_file
        self._override_system_prompt = system_prompt  # None = 使用 yaml

    # ── 属性懒加载 ──────────────────────────────────────────

    @property
    def _cfg(self):
        """懒加载 PromptsConfig（带缓存）"""
        from .config_loader import load_prompts_config
        return load_prompts_config(self._prompts_file)

    @property
    def system_prompt(self) -> str:
        return self._override_system_prompt or self._cfg.system_prompt

    @system_prompt.setter
    def system_prompt(self, value: str):
        self._override_system_prompt = value

    # ── 公开接口 ────────────────────────────────────────────

    def build(
        self,
        query: str,
        context_blocks: List[ContextBlock],
        context_truncated: bool = False,
    ) -> Tuple[str, str]:
        """
        构建完整的推理提示词
        Returns: (system_prompt, user_prompt)
        """
        context = self._format_context(context_blocks)

        extended_system = self.system_prompt
        if context_truncated:
            extended_system += self._cfg.system_truncation_suffix

        user_prompt = (
            self._cfg.user_prompt_template
            .replace('{context}', context)
            .replace('{query}', query)
        )

        return extended_system, user_prompt

    def build_rejection(self, max_score: float) -> Tuple[str, str]:
        """
        构建拒答提示词
        Returns: (system_prompt, user_prompt)
        """
        user = self._cfg.rejection_prompt.replace('{max_score}', f'{max_score:.2f}')
        return self.system_prompt, user

    def build_no_llm(
        self,
        query: str,
        context_blocks: List[ContextBlock],
    ) -> Tuple[str, str]:
        """
        构建无 LLM 模式提示词
        Returns: (system_prompt, user_prompt)
        """
        context = self._format_context(context_blocks)
        summary = self._summarize_results(context_blocks)
        user = (
            self._cfg.no_llm_prompt
            .replace('{query}', query)
            .replace('{context}', context)
            .replace('{summary}', summary)
        )
        return self._cfg.no_llm_system_prompt, user

    def extract_citation_ids(self, answer: str) -> List[int]:
        """
        从 LLM 回答中提取引用 ID
        使用 dict.fromkeys 保序去重，避免 O(n²) 的 list + not in 模式。
        """
        raw = [int(m.group(1)) for m in re.finditer(r'\[(\d+)\]', answer)]
        return list(dict.fromkeys(raw))

    def build_stream_message(
        self,
        query: str,
        context_blocks: List[ContextBlock],
        context_truncated: bool = False,
    ) -> str:
        """
        构建用于流式生成的单条消息
        """
        context = self._format_context(context_blocks)
        warning = self._cfg.stream_truncation_warning if context_truncated else ''

        return (
            self._cfg.stream_message_template
            .replace('{context}', context)
            .replace('{truncation_warning}', warning)
            .replace('{query}', query)
        )

    # ── 私有辅助 ────────────────────────────────────────────

    def _format_context(self, blocks: List[ContextBlock]) -> str:
        """格式化上下文块"""
        if not blocks:
            return '（无相关文档）'
        parts = []
        for block in blocks:
            formatted = f"[ID: {block.id}, Source: {block.source}]\n{block.content}"
            if block.is_truncated:
                formatted += '\n[此段内容已截断，建议查阅原文]'
            parts.append(formatted)
        return '\n\n---\n\n'.join(parts)

    def _summarize_results(self, blocks: List[ContextBlock]) -> str:
        """汇总检索结果（无 LLM 模式）"""
        lines = []
        for i, block in enumerate(blocks):
            snippet = block.content[:200]
            suffix = '...' if len(block.content) > 200 else ''
            lines.append(f"{i + 1}. {snippet}{suffix}")
        return '\n'.join(lines)


def create_prompt_builder(
    system_prompt: Optional[str] = None,
    prompts_file: Optional[str] = None,
) -> PromptBuilder:
    """创建提示词构建器"""
    return PromptBuilder(system_prompt, prompts_file)
